File: src/main/java/main/json/JsonParser.java
package main.json;

import main.App;
import main.puzzle.Shape;
import main.puzzle.*;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 *
 * @author Bunnyspa
 */
public class JsonParser {

    private static final String SIGNKEY = "sign";
    private static final String CHIPKEY_SQUAD = "squad_with_user_info";
    private static final String CHIPKEY_CHIP = "chip_with_user_info";
    private static final String UNKNOWN_HOC = "Unknown HOC";

    @NotNull
    public static List<Chip> readFile(@NotNull String filePath) {
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String data = br.lines().collect(Collectors.joining());
            List<Chip> chips = parseChip(data);
            return chips;
        } catch (IOException ex) {
            App.log(ex);
            return new ArrayList<>();
        }
    }

    public static String parseSign(@NotNull String data) {
        ObjectJson o = Json.getObjectJson(Json.parse(data));
        return Json.getText(o.getValue(SIGNKEY));
    }

    @NotNull
    public static List<Chip> parseChip(@NotNull String data) {
        // Init
        Map<Integer, Tag> squadMap = new HashMap<>();
        List<Chip> chips = new ArrayList<>();
        try {
            // Parse
            ObjectJson o = Json.getObjectJson(Json.parse(data));

            // Squad
            Json squadsJ = o.getValue(CHIPKEY_SQUAD);
            if (squadsJ.getType() == Json.OBJECT) {
                List<String> squadJKeys = Json.getObjectKeys(squadsJ);
                parseChip_squad(squadMap, squadJKeys.stream().map((jKey) -> Json.getObjectJson(Json.getObjectValue(squadsJ, jKey))));
            } else if (squadsJ.getType() == Json.ARRAY) {
                List<Json> squadJs = Json.getList(squadsJ);
                parseChip_squad(squadMap, squadJs.stream().map(Json::getObjectJson));
            }

            // Chip
            Json chipsJ = o.getValue(CHIPKEY_CHIP);
            if (chipsJ.getType() == Json.OBJECT) {
                List<String> chipJKeys = Json.getObjectKeys(chipsJ);
                parseChip_chip(squadMap, chips, chipJKeys.stream().map((jKey) -> Json.getObjectJson(Json.getObjectValue(chipsJ, jKey))));
            } else if (chipsJ.getType() == Json.ARRAY) {
                List<Json> chipJs = Json.getList(chipsJ);
                parseChip_chip(squadMap, chips, chipJs.stream().map(Json::getObjectJson));
            }

            Collections.reverse(chips);
        } catch (Exception ignored) {
        }
        return chips;
    }

    private static void parseChip_squad(@NotNull Map<Integer, Tag> squadMap, @NotNull Stream<ObjectJson> stream) {
        stream.forEach((squadJ) -> {
            int squadID = Integer.parseInt(Json.getText(squadJ.getValue("id")));
            int squadIndex = Integer.parseInt(Json.getText(squadJ.getValue("squad_id"))) - 1;
            squadMap.put(squadID, boardTag(squadIndex));
        });
    }

    private static void parseChip_chip(@NotNull Map<Integer, Tag> squadMap, @NotNull List<Chip> chips, @NotNull Stream<ObjectJson> stream) {
        stream.forEach((chipJ) -> {
            // Raw
            String id = ((TextJson) chipJ.getValue("id")).getText();
            int gridData = Integer.parseInt(((TextJson) chipJ.getValue("grid_id")).getText());
            Shape shape = Shape.byId(gridData);
            int dmg = Integer.parseInt(Json.getText(chipJ.getValue("assist_damage")));
            int brk = Integer.parseInt(Json.getText(chipJ.getValue("assist_def_break")));
            int hit = Integer.parseInt(Json.getText(chipJ.getValue("assist_hit")));
            int rld = Integer.parseInt(Json.getText(chipJ.getValue("assist_reload")));
            int star = Integer.parseInt(Json.getText(chipJ.getValue("chip_id")).substring(0, 1));
            int level = Integer.parseInt(Json.getText(chipJ.getValue("chip_level")));
            int color = Integer.parseInt(Json.getText(chipJ.getValue("color_id"))) - 1;
            int rotation = Integer.parseInt(Json.getText(chipJ.getValue("shape_info")).substring(0, 1));
            int squadID = Integer.parseInt(Json.getText(chipJ.getValue("squad_with_user_id")));

            Stat pt = new Stat(dmg, brk, hit, rld);
            Chip chip = new Chip(id, shape, star, color, pt, level, rotation);
            if (squadMap.containsKey(squadID)) {
                Tag tag = squadMap.get(squadID);
                chip.setTag(tag, true);
            } else if (squadID != 0) {
                Tag tag = boardTag(squadID - 10001);
                chip.setTag(tag, true);
            }
            chips.add(chip);
        });
    }

    @NotNull
    private static Tag boardTag(int index) {
        if (index < Board.NAMES.length) {
            return new Tag(Color.GRAY, Board.NAMES[index]);
        }
        return new Tag(Color.GRAY, UNKNOWN_HOC);
    }
}

File: src/main/java/main/json/JsonParser.kt
package main.json

import main.App
import main.puzzle.*
import java.awt.Color
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

/**
 *
 * @author Bunnyspa
 */
object JsonParser {
    private const val SIGNKEY = "sign"
    private const val CHIPKEY_SQUAD = "squad_with_user_info"
    private const val CHIPKEY_CHIP = "chip_with_user_info"
    private const val UNKNOWN_HOC = "Unknown HOC"
    fun readFile(filePath: String): List<Chip> {
        try {
            BufferedReader(FileReader(filePath)).use { br ->
                val data = br.lines().collect(Collectors.joining())
                return parseChip(data)
            }
        } catch (ex: IOException) {
            App.log(ex)
            return mutableListOf()
        }
    }

    fun parseSign(data: String): String {
        val o: ObjectJson = Json.getObjectJson(Json.parse(data))
        return Json.getText(o.getValue(SIGNKEY)!!)
    }

    fun parseChip(data: String): List<Chip> {
        // Init
        val squadMap: MutableMap<Int, Tag> = HashMap()
        val chips: MutableList<Chip> = mutableListOf()
        try {
            // Parse
            val o: ObjectJson = Json.getObjectJson(Json.parse(data))

            // Squad
            val squadsJ = o.getValue(CHIPKEY_SQUAD)
            if (squadsJ != null) {
                if (squadsJ.type == Json.OBJECT) {
                    val squadJKeys = Json.getObjectKeys(squadsJ)
                    parseChip_squad(squadMap, squadJKeys.stream().map { jKey: String? ->
                        Json.getObjectJson(
                            Json.getObjectValue(squadsJ, jKey)!!
                        )
                    })
                } else if (squadsJ.type == Json.ARRAY) {
                    val squadJs = Json.getList(squadsJ)
                    parseChip_squad(
                        squadMap,
                        squadJs.stream().map { data: Json? -> Json.getObjectJson(data) }
                    )
                }
            }

            // Chip
            val chipsJ = o.getValue(CHIPKEY_CHIP)
            if (chipsJ?.type == Json.OBJECT) {
                val chipJKeys: List<String?> = Json.getObjectKeys(chipsJ)
                parseChip_chip(squadMap, chips, chipJKeys.stream().map { jKey: String? ->
                    Json.getObjectJson(
                        Json.getObjectValue(chipsJ, jKey)!!
                    )
                })
            } else if (chipsJ?.type == Json.ARRAY) {
                val chipJs: List<Json?> = Json.getList(chipsJ)
                parseChip_chip(
                    squadMap,
                    chips,
                    chipJs.stream().map { data: Json? -> Json.getObjectJson(data) }
                )
            }
            chips.reverse()
        } catch (ignored: Exception) {
        }
        return chips
    }

    private fun parseChip_squad(squadMap: MutableMap<Int, Tag>, stream: Stream<ObjectJson>) {
        stream.forEach { squadJ: ObjectJson ->
            val squadID: Int = Json.getText(squadJ.getValue("id")!!).toInt()
            val squadIndex: Int = Json.getText(squadJ.getValue("squad_id")!!).toInt() - 1
            squadMap[squadID] = boardTag(squadIndex)
        }
    }

    private fun parseChip_chip(squadMap: Map<Int, Tag>, chips: MutableList<Chip>, stream: Stream<ObjectJson>) {
        stream.forEach { chipJ: ObjectJson ->
            // Raw
            val id = (chipJ.getValue("id") as TextJson?)?.text
            val gridData = (chipJ.getValue("grid_id") as TextJson).text.toInt()
            val shape = Shape.byId(gridData)
            val dmg: Int = Json.getText(chipJ.getValue("assist_damage")!!).toInt()
            val brk: Int = Json.getText(chipJ.getValue("assist_def_break")!!).toInt()
            val hit: Int = Json.getText(chipJ.getValue("assist_hit")!!).toInt()
            val rld: Int = Json.getText(chipJ.getValue("assist_reload")!!).toInt()
            val star: Int = Json.getText(chipJ.getValue("chip_id")!!).substring(0, 1).toInt()
            val level: Int = Json.getText(chipJ.getValue("chip_level")!!).toInt()
            val color: Int = Json.getText(chipJ.getValue("color_id")!!).toInt() - 1
            val rotation: Int = Json.getText(chipJ.getValue("shape_info")!!).substring(0, 1).toInt()
            val squadID: Int = Json.getText(chipJ.getValue("squad_with_user_id")!!).toInt()
            val pt = Stat(dmg, brk, hit, rld)
            val chip = Chip(id, shape, star, color, pt, level, rotation)
            if (squadMap.containsKey(squadID)) {
                val tag = squadMap.getValue(squadID)
                chip.setTag(tag, true)
            } else if (squadID != 0) {
                val tag = boardTag(squadID - 10001)
                chip.setTag(tag, true)
            }
            chips.add(chip)
        }
    }

    private fun boardTag(index: Int): Tag {
        return if (index < Board.NAMES.size) {
            Tag(Color.GRAY, Board.NAMES[index])
        } else Tag(
            Color.GRAY,
            UNKNOWN_HOC
        )
    }
}